src/internal/drawImage.js

```diff
@@ -6,9 +18,169 @@
  * @returns {void}
  * @memberof Internal
  */
-export default function (enabledElement, invalidated = false) {
-  enabledElement.needsRedraw = true;
+
+
+export function drawImage (enabledElement, invalidated = false) {
+  registerVisibilityChange();
+
+  if (!enabledElement || !enabledElement.canvas || (!enabledElement.image && enabledElement.layers.length < 1)) {
+    return;
+  }
+  
+  const element = enabledElement.element;
+  
   if (invalidated) {
     enabledElement.invalid = true;
   }
+
+  if (!elementsDrawMap.has(element)) {
+    elementsDrawMap.set(element, { 
+      needsRedraw: false, 
+      callbacks: new Set()
+    });
+  }
+
+  elementsDrawMap.get(element).needsRedraw = true;
+
+  rafStopTime = performance.now() + rafStopMaxTime;
+  if (!rafId) {
+    rafId = requestAnimationFrame(step);
+  }
+}
+
+function step () {
+  elementsDrawMap.forEach((entry, element) => {
+    if (!element || !entry || !entry.needsRedraw) {
+      return;
+    }
+    
+    const hasCallbacks = entry.callbacks.size > 0;
+    
+    entry.needsRedraw = hasCallbacks;
+
+    let enabledElement;
+
+    try { 
+      enabledElement = getEnabledElement(element); 
+    } catch (err) { 
+      return; 
+    }
+    
+    const eventDetails = {
+      enabledElement,
+      timestamp: Date.now()
+    };
+    
+    triggerEvent(element, EVENTS.PRE_RENDER, eventDetails);
+    
+    drawImageSync(enabledElement, enabledElement.invalid);
+
+    if (hasCallbacks) {
+      entry.callbacks.forEach((rafcb) => {
+        try {
+          rafcb(enabledElement);
+        } catch (err) {
+          // swallow
+        }
+      });
+      rafStopTime = performance.now() + rafStopMaxTime; 
+    }
+  });
+  
+  if (rafStopTime && performance.now() < rafStopTime) {
+    rafId = requestAnimationFrame(step);
+  } else {
+    stopRaf();
+  }
+}
+
+/**
+ * add a callback to be called on every RequestAnimationFrame step
+ *
+ * @param {HTMLElement} element The Cornerstone element for this callback
+ * @param {Function} f The callback to call, will be called with the enabled cornerstone element
+ * @returns {void}
+ * @memberof Internal
+ */
+export function addDrawCallback (element, f) {
+  if (!element) {
+    return;
+  }
+  if (!elementsDrawMap.has(element)) {
+    elementsDrawMap.set(element, { 
+      needsRedraw: false, 
+      callbacks: new Set()
+    });
+  }
+
+  elementsDrawMap.get(element).callbacks.add(f);
+}
+
+
+/**
+ * delete a callback that was added with addDrawCallback
+ *
+ * @param {HTMLElement} element The Cornerstone element for this callback
+ * @param {Function} f The callback to delete
+ * @returns {void}
+ * @memberof Internal
+ */
+export function removeDrawCallback (element, f) {
+  if (!element) {
+    return;
+  }
+  const entry = elementsDrawMap.get(element);
+  
+  if (!entry) {
+    return;
+  }
+  entry.callbacks.delete(f);
+}
+
+
+function registerVisibilityChange () {
```

Comment:
Needs some comment.
If I understand correctly, this is to stop RAF when the browser tab becomes hidden? 
I was under the impression that modern browser already do that? Is this really required?

Reply:
This is more to restart RAF when the tab regains visibility

It is is true that RAF usually stops automatically when visibility is lost but I understood this is not 100% reliable so I manually stop it as well just to be sure (by calling `stopRAF`)